drivers/gpio/gpio_sam0.c

```diff
@@ -172,22 +117,151 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
 			  u32_t *value)
 {
 	const struct gpio_sam0_config *config = DEV_CFG(dev);
-	u32_t bits;
 
-	if (access_op != GPIO_ACCESS_BY_PIN) {
-		/* TODO(mlhx): support GPIO_ACCESS_BY_PORT */
-		return -ENOTSUP;
-	}
+	*value = (config->regs->IN.reg & BIT(pin)) != 0;
 
-	bits = config->regs->IN.bit.IN;
-	*value = (bits >> pin) & 1;
+	return 0;
+}
+
+static int gpio_sam0_port_get_raw(struct device *dev,
+				  gpio_port_value_t *value)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+
+	*value = config->regs->IN.reg;
+
+	return 0;
+}
+
+static int gpio_sam0_port_set_masked_raw(struct device *dev,
+					 gpio_port_pins_t mask,
+					 gpio_port_value_t value)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+	u32_t out = config->regs->OUT.reg;
+
+	config->regs->OUT.reg = (out & ~mask) | (value & mask);
+
+	return 0;
+}
+
+static int gpio_sam0_port_set_bits_raw(struct device *dev,
+				       gpio_port_pins_t pins)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+
+	config->regs->OUTSET.reg = pins;
+
+	return 0;
+}
+
+static int gpio_sam0_port_clear_bits_raw(struct device *dev,
+					 gpio_port_pins_t pins)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+
+	config->regs->OUTCLR.reg = pins;
+
+	return 0;
+}
+
+static int gpio_sam0_port_toggle_bits(struct device *dev,
+				      gpio_port_pins_t pins)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+
+	config->regs->OUTTGL.reg = pins;
 
 	return 0;
 }
 
 #ifdef CONFIG_SAM0_EIC
 
-int gpio_sam0_manage_callback(struct device *dev,
+static int gpio_sam0_pin_interrupt_configure(struct device *dev,
+					     unsigned int pin,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
+{
+	const struct gpio_sam0_config *config = DEV_CFG(dev);
+	PortGroup *regs = config->regs;
+	PORT_PINCFG_Type pincfg = {
+		.reg = regs->PINCFG[pin].reg,
+	};
+	enum sam0_eic_trigger trigger;
+	int rc = 0;
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		pincfg.bit.PMUXEN = 0;
+		rc = sam0_eic_disable_interrupt(config->id, pin);
+		if (rc == -EBUSY) {
+			/* Ignore diagnostic disabling disabled */
+			rc = 0;
+		}
+		if (rc == 0) {
+			rc = sam0_eic_release(config->id, pin);
+		}
+		break;
+	case GPIO_INT_MODE_LEVEL:
+	case GPIO_INT_MODE_EDGE:
+		/* Interrupt detection requires input buffer */
+		if (pincfg.bit.INEN == 0) {
+			rc = -EINVAL;
```

Comment:
We should return `-ENOTSUP` since the configuration is valid, it's however not supported by the hardware. That's also what other drivers are doing. With `-EINVAL` the 1 pin testcase will not pass.